### PR TITLE
Add a Snowplow interface with the ability to initialize and manage multiple trackers (close #22)

### LINKED ISSUE #22: Add a Snowplow interface with the ability to initialize and manage multiple trackers
Hi! I work for a company that makes and sells audio plugins, and we're looking at different solutions for adding tracking to our apps. 

The thing with audio plugins is that they are (usually) dynamically loaded into DAWs ("digital audio workstations", the audio plugin hosts), and it is a very common occurance to have a single audio plugin loaded multiple times in the same process (for instance: if you have an audio plugin that works as some kind of filter or audio effect, you may want that filter on several different tracks). 

Snowplow seems like a very nice solution for this, except for one thing: the "global singleton" nature of the tracker. That's not ideal for us, it would be really nice to load one tracker for each plugin, if possible. 

I was looking through the code an seeing if that was possible: seemingly it is: if you create multiple trackers with `Tracker::init`, the first one gets stored as the global singleton, but it's not an error to create multiple ones, you just have destroy it yourself (and not use `Tracker::instance`). I realize that the Storage class is also a global singleton, but I figure that's less of an issue, since it's locked with a mutex, and I plan on passing the same database path for every instance. 

However, when implementing this solution, the compiler informed me of another problem: the Tracker destructor is private, so you literally *can't* manage it yourself: the only way to destroy a Tracker is through `Tracker::close()`, which destroys the singleton. 

One solution to this is simply making the Tracker destructor (and constructor as well, while we're at it) public, and let me totally ignore the singleton functions, they serve no purpose for me anyway. However, at this point, I would like to get advice from you if this would be a good change. Is there any reason you can think of why this would be a problem? 

I understand you wanting to offer a singleton as a convenience, but forcing its use like you do now is very thing for a library to do: for many kinds of applications, it's a very dangerous thing to use. There's a reason Singletons are considered an anti-pattern by most authorities.

Incidentally: if you are going to force the use of a global singleton, there are a few fixes that should probably be done: 

 1. Tracker::init should throw an exception if the singleton already exists (since the only way to delete the class is by deleting the singleton, it should be an error to create any other object). 
 2. The singleton should not have the type `Tracker*`, it should be a `std::unique_ptr<Tracker>`, to ensure that it is always uninitialized.

### PR DESCRIPTION
**Pull request by matus-tomlein**
This PR addresses issue #22 which goal was to enable multiple tracker instances. To do so, it takes a larger refactoring of the API for intializing trackers. It does the following:

1. It removes the tracker singleton instance and `Tracker::init()` and `Tracker::instance()` and replaces them with static API in a new `Snowplow` static class (which largely reflects the mobile trackers API):
    * `Snowplow::create_tracker(...)` to create a new tracker instance.
    * `Snowplow::get_tracker(ns)` to retrieve a tracker instance by it's namespace.
    * `Snowplow::get_default_tracker()` to get the default tracker instance (first one or the one set by `Snowplow::set_default_tracker(tracker)`.
    * `Snowplow::remove_tracker(tracker)` to remove reference to a tracker instance.
2. It adds configuration classes (`TrackerConfiguration`, `NetworkConfiguration`, `EmitterConfiguration`, `SessionConfiguration`) used to initialize new trackers through `Snowplow::create_tracker()` (again inspired by the mobile trackers API).
3. It also refactors the constructors for `Tracker`, `Emitter`, `ClientSession` (which can still be used directly without the `Snowplow` interface) to add default values, remove unnecessary raw pointers, and adopt C++ smart pointers (e.g., `shared_ptr`) where necessary.

Thanks to the default values and `Snowplow::create_tracker` call, it is now much easier to instantiate a new tracker and takes much less code (can be done in 1 line of code). I described the initialisation in the docs which I paste below. Also, [here is a PR](https://github.com/snowplow-incubator/data-value-resources/pull/104) for the documentation change in data-value-resources.

# Documentation

There are 3 ways to initialize a tracker instance that you will use to track events:

1. Create a tracker with default configuration using the `Snowplow` interface.
3. Create a tracker with custom configuration using the `Snowplow` interface.
5. Create and manage `Tracker`, `Emitter`, and `ClientSession` directly.

## Option 1: Creating a tracker with default configuration using the `Snowplow` interface[](#creating-a-tracker)

The `Snowplow` class provides static methods that let you easily create a new tracker. It can be as simple as:

```cpp
auto tracker = Snowplow::create_tracker(
    "ns", // tracker namespace used to identify the tracker
    "https://com.acme.collector", // Snowplow collector URL
    POST, // HTTP method used to send events to the collector
    "events.db" // Relative path to an SQLite database used for event queue and session tracking
);
```

Optionally, you may choose to attach a `Subject` instance with information about the user and device (see the next page on [Adding data](03-adding-data.md) to learn more about the `Subject`):

```cpp
auto subject = std::make_shared<Subject>() // initialize a C++ shared pointer for the Subject
subject->set_user_id("a-user-id");

auto tracker = Snowplow::create_tracker("ns", "https://com.acme.collector", POST, "events.db", subject);
```

Finally, client session tracking will be automatically enabled and session information will be attached to all events (see documentation on [Client sessions](06-client-sessions.md) for more details). If you wish to disable session tracking, pass `false` as the last argument:

```cpp
auto tracker = Snowplow::create_tracker("ns", "https://com.acme.collector", POST, "events.db", subject, false);
```

After you create a tracker, you can access it from anywhere using it's namespace:

```cpp
auto tracker = Snowplow::get_tracker("ns");
```

You can also access the default tracker using `Snowplow::get_default_tracker()`. The default tracker is the first initialized tracker (unless it is removed). If you have multiple trackers, you can choose the default tracker by calling `Snowplow::set_default_tracker(tracker2)`.

To remove reference of previously initialized tracker from the Snowplow interface (the tracker will be deleted once all remaininig references to it are deleted):

```cpp
Snowplow::remove_tracker(tracker);
```

## Option 2: Creating a tracker with custom configuration using the `Snowplow` interface

There are a number of settings that you may want to customize when creating a tracker instance. You can make use of configuration objects to provide custom configuration for the tracker:

```cpp
TrackerConfiguration tracker_config(
    "namespace", // tracker namespace
    "app-id", // application ID
    "mob" // platform that the tracker runs on
);
tracker_config.set_desktop_context(false); // do not attach a desktop context entity to events (on by default)

NetworkConfiguration network_config("https://com.acme.collector", POST); // collector URL and method

EmitterConfiguration emitter_config("sp.db"); // event queue DB path if SQLite or custom `EventStore` instance
emitter_config.set_batch_size(500); // maximum number of events to send at a time

SessionConfiguration session_config(
    "sp.db", // session DB path if SQLite (usually same as event queue) or custom `SessionStore` instance
    5000, // foreground timeout in ms
    5000 // background timeout in ms
);

auto tracker = Snowplow::create_tracker(tracker_config, network_config, emitter_config, session_config);
```

### Tracker configuration using `TrackerConfiguration`

`TrackerConfiguration` contains settings to identify and configure the tracker. It's constructor takes 3 attributes:

| Constructor attribute | Description | Default |
|---|---|---|
| `name_space` | Tracker namespace to identify the tracker and also attach as a property to tracked events. | None |
| `app_id` | Application ID. | "" |
| `platform` | The platform the Tracker is running on, can be one of: web, mob, pc, app, srv, tv, cnsl, iot | "srv" |

It further provides 2 optional setter functions:

| Setter | Description | Default |
|---|---|---|
| `set_use_base64` | Whether to use base64 encoding in events. | true |
| `set_desktop_context` | Whether to add a desktop_context, which gathers information about the device the tracker is running on, to each event. | true |

### Network configuration using `NetworkConfiguration`

`NetworkConfiguration` has only two properties set in it's constructor to configure the Snowplow collector:

| Constructor attribute | Description | Default |
|---|---|---|
| `collector_url` | Full URL of the Snowplow collector including the protocol (or defaults to HTTPS if protocol not present). | None |
| `method` | HTTP method to use when sending events to collector – GET or POST. | POST |

### Emitter configuration using `EmitterConfiguration`

`EmitterConfiguration` brings additional settings for the constructor. It provides two constructors that accept the event store either as a path to the SQLite database or a custom `EventStore` object. The following configurations are identical:

```cpp
EmitterConfiguration config1("sp.db");

auto storage = std::make_shared<SqliteStorage>("sp.db");
EmitterConfiguration config2(storage); // you can also pass a custom `EventStore`
```

Additionally, it provides the following setter functions:

| Setter | Description | Default |
|---|---|---|
| `set_batch_size` | The maximum amount of events to send at a time. | 250 events |
| `set_byte_limit_get` | The byte limit when sending a GET request. | 40000 bytes |
| `set_byte_limit_post` | The byte limit when sending a POST request. | 40000 bytes |
| `set_http_client` | Unique pointer to a custom HTTP client to send GET and POST requests with. | Platform-specific implementation. |
| `set_request_callback` | Set a callback to call after emit requests are made with the resulting emit status (see page about Emitter for more info). | None |
| `set_custom_retry_for_status_code` | Set a custom retry rule for when the HTTP status code is received in emit response from Collector (see page about Emitter for more details). | None |

### Session configuration using `SessionConfiguration`

`SessionConfiguration` is an optional attribute in `create_tracker`. If passed, session tracking will be enabled with the given configuration. If not passed, session tracking will not be enabled. The constructor takes 3 attributes:

| Constructor attribute | Description | Default |
|---|---|---|
| `session_store` or `db_name` | You may either pass a path to an SQLite database to be used for session storage or a custom implementation of `SessionStore` similar as for `EventStore` in `EmitterConfiguration`. | None |
| `foreground_timeout` | Timeout in ms for updating the session when the app is in background. | 30 minutes |
| `background_timeout` | Timeout in ms for updating the session when the app is in foreground. | 30 minutes |

## Option 3: Managing `Tracker`, `Emitter`, and `ClientSession` directly

The third option to initialise a new tracker is to instantiate it and the related components directly. This option is suitable in case you want supply a custom emitter or client session implementation. If you don't want to do that, we recommend using Option 2 which gives you the same configuration options with a simpler API.

The following example takes you through the steps needed to initialize a storage, emitter, subject, client session, and tracker:

```cpp
#include "snowplow.hpp"

// 1. create storage for event queue and session information
auto storage = std::make_shared<SqliteStorage>("sp.db");
// 2. create an emitter for sending event to Snowplow collector
auto emitter = std::make_shared<Emitter>(storage, "com.acme.collector");
// 3. create a subject with user and device information
auto subject = std::make_shared<Subject>();
subject->set_user_id("a-user-id");
// 4. optionally create a client_session for session tracking
auto client_session = std::make_shared<ClientSession>(storage, 5000, 5000);
// 5. finally, create the tracker instance
auto tracker = std::make_shared<Tracker>(emitter, subject, client_session, "pc", "app_id", "ns");
```

Once you initialize the tracker, you can optionally register it with the Snowplow interface:

```cpp
Snowplow::register_tracker(tracker);
```

This will make accessible from anywhere using `Snowplow::get_tracker("ns")`.

The following will explain the individual components shown in the above code sample in more detail.

### Storage

The storage has two functions in the example above – it is used by the `emitter` to persist an event queue with events to be sent, and it is used by the `client_session` to persist the current session. The tracker provides an SQL storage (`SqliteStorage`) implementation, but you may introduce your own storage as described in ["Emitters"](05-emitters.md) and ["Client Sessions"](06-client-sessions.md).

### Emitter

Emitter is a required component responsible for sending tracked events to the collector.

Accepts an argument of an Emitter instance pointer; if the object is `NULL` will throw an exception. See Emitters for more on emitter configuration.

| **Argument Name** | **Description** | **Required?** | **Default** |
| --- | --- | --- | --- |
| `event_store` | Defines the database to use for event queue | Yes | – |
| `uri` | The collector URI (excluding protocol) to send events to | Yes | – |
| `method` | The request type to use (GET or POST) | No | POST |
| `protocol` | The protocol to use (HTTP or HTTPS) | No | HTTPS |
| `batch_size` | The maximum amount of events to send at a time | No | 250 events |
| `byte_limit_post` | The byte limit when sending a POST request | No | 40000 bytes |
| `byte_limit_get` | The byte limit when sending a GET request | No | 40000 bytes |
| `http_client` | Unique pointer to a custom HTTP client to send GET and POST requests with | No | Platform-specific implementation. |

### Subject

The user which the Tracker will track. Accepts an argument of a Subject instance pointer.

You don’t need to set this during Tracker construction; you can use the `tracker.set_subject(...)` method afterwards. In fact, you don’t need to create a subject at all. If you don’t, though, your events won’t contain user-specific data such as timezone and language.

### Client session

The client sessions which the Tracker will attach to each event. Accepts an argument of a ClientSession instance pointer.

Adds the ability to attach a ClientSession context to each event that leaves the Tracker. This object will persistently store information about the sessions that have occurred for the life of the application – unless the database is destroyed.

### Tracker

The tracker instance lets you track events. You should maintain a reference to the initialized tracker during the app lifetime.

| **Argument Name** | **Description** | **Required?** | **Default** |
| --- | --- | --- | --- |
| `emitter` | The emitter to which events are sent | Yes | – |
| `subject` | The user being tracked | No  | nullptr |
| `client_session` | Client session recording | No  | nullptr |
| `platform` | The platform the Tracker is running on | No  | "srv" |
| `app_id` | The application ID | No  | "" |
| `name_space` | The namespace of the tracker instance used to identify the tracker. | No  | "" |
| `use_base64` | Whether to enable [base 64 encoding](https://en.wikipedia.org/wiki/Base64). Defaults to true to ensure that no data is lost or corrupted. | No  | true |
| `desktop_context` | Whether to add a `desktop_context` to events | No  | true |

The `desktop_context` gathers extra information about the device it is running on and sends it along with every event that is made by the Tracker.

An example of the data in this context:

```json
{
    "deviceManufacturer": "Apple Inc.",
    "deviceModel": "MacPro3,1",
    "deviceProcessorCount": 8,
    "osIs64Bit": true,
    "osServicePack": "",
    "osType": "macOS",
    "osVersion": "10.11.2"
}
```

For more information the raw JsonSchema can be found [here](https://raw.githubusercontent.com/snowplow/iglu-central/master/schemas/com.snowplowanalytics.snowplow/desktop_context/jsonschema/1-0-0).

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

File: src/constants.hpp [hunk 1]
```diff
@@ -101,6 +101,17 @@ const string SNOWPLOW_SESSION_PREVIOUS_ID = "previousSessionId";
...
+const unsigned long long SNOWPLOW_SESSION_DEFAULT_TIMEOUT = 30 * 1000 * 1000; // 30 minutes
+
+// emitter defaults
+const int SNOWPLOW_EMITTER_DEFAULT_BATCH_SIZE = 250;
+const int SNOWPLOW_EMITTER_DEFAULT_BYTE_LIMIT_GET = 40000;
+const int SNOWPLOW_EMITTER_DEFAULT_BYTE_LIMIT_POST = 40000;
+
+// tracker defaults
+const string SNOWPLOW_DEFAULT_APP_ID = "";
+const string SNOWPLOW_DEFAULT_PLATFORM = "srv";
```
Comment (AlexBenny) [inline src/constants.hpp:113]: I guess the "platform" is an enum in the schema. Perhaps, we can create an enum with the constants here too?
would it be clumsy?

File: examples/main.cpp [hunk 2]
```diff
@@ -85,18 +94,17 @@ int main(int argc, char **argv) {
     se.property = &property;
     se.value = &value;
 
-    t->track(te);
-    t->track(sve);
-    t->track(se);
+    tracker->track(te);
+    tracker->track(sve);
+    tracker->track(se);
```
Comment (AlexBenny) [inline examples/main.cpp:99]: 🙏

**Review by AlexBenny — APPROVED** [review]
LGTM 👍 Lot of work!

File: src/constants.hpp (hunk 1 above)
Comment (matus-tomlein) [reply src/constants.hpp:113]: Good idea, changed the platform to an enum in the TrackerConfiguration!

**Pull request merged**